core/services/llo/offchain_config_digester.go

```diff
@@ -0,0 +1,126 @@
+
+func NewOffchainConfigDigester(chainID *big.Int, contractAddress common.Address) OffchainConfigDigester {
+	return OffchainConfigDigester{chainID, contractAddress}
+}
+
+type OffchainConfigDigester struct {
+	ChainID         *big.Int
+	ContractAddress common.Address
+}
+
+func (d OffchainConfigDigester) ConfigDigest(cc ocrtypes.ContractConfig) (ocrtypes.ConfigDigest, error) {
+	signers := []common.Address{}
+	for i, signer := range cc.Signers {
+		if len(signer) != 20 {
+			return ocrtypes.ConfigDigest{}, errors.Errorf("%v-th evm signer should be a 20 byte address, but got %x", i, signer)
+		}
+		a := common.BytesToAddress(signer)
+		signers = append(signers, a)
+	}
+	transmitters := []credentials.StaticSizedPublicKey{}
+	for i, transmitter := range cc.Transmitters {
+		if len(transmitter) != 2*ed25519.PublicKeySize {
+			return ocrtypes.ConfigDigest{}, errors.Errorf("%v-th evm transmitter should be a 64 character hex-encoded ed25519 public key, but got '%v' (%d chars)", i, transmitter, len(transmitter))
+		}
+		var t credentials.StaticSizedPublicKey
+		b, err := hex.DecodeString(string(transmitter))
+		if err != nil {
+			return ocrtypes.ConfigDigest{}, errors.Wrapf(err, "%v-th evm transmitter is not valid hex, got: %q", i, transmitter)
+		}
+		copy(t[:], b)
+
+		transmitters = append(transmitters, t)
+	}
+
+	return configDigest(
+		d.ChainID,
+		d.ContractAddress,
+		cc.ConfigCount,
+		signers,
+		transmitters,
+		cc.F,
+		cc.OnchainConfig,
+		cc.OffchainConfigVersion,
+		cc.OffchainConfig,
+	), nil
+}
+
+func (d OffchainConfigDigester) ConfigDigestPrefix() (ocrtypes.ConfigDigestPrefix, error) {
+	return ocrtypes.ConfigDigestPrefixLLO, nil
+}
+
+func makeConfigDigestArgs() abi.Arguments {
+	abi, err := abi.JSON(strings.NewReader(exposed_channel_verifier.ExposedChannelVerifierABI))
+	if err != nil {
+		// assertion
+		panic(fmt.Sprintf("could not parse aggregator ABI: %s", err.Error()))
+	}
+	return abi.Methods["exposedConfigDigestFromConfigData"].Inputs
+}
+
+var configDigestArgs = makeConfigDigestArgs()
+
+func configDigest(
+	chainID *big.Int,
+	contractAddress common.Address,
+	configCount uint64,
+	oracles []common.Address,
+	transmitters []credentials.StaticSizedPublicKey,
+	f uint8,
+	onchainConfig []byte,
+	offchainConfigVersion uint64,
+	offchainConfig []byte,
+) types.ConfigDigest {
+	msg, err := configDigestArgs.Pack(
+		chainID,
+		contractAddress,
+		configCount,
+		oracles,
+		transmitters,
+		f,
+		onchainConfig,
+		offchainConfigVersion,
+		offchainConfig,
+	)
+	if err != nil {
+		// assertion
+		panic(err)
+	}
+	rawHash := crypto.Keccak256(msg)
+	configDigest := types.ConfigDigest{}
+	if n := copy(configDigest[:], rawHash); n != len(configDigest) {
+		// assertion
+		panic("copy too little data")
+	}
+	binary.BigEndian.PutUint16(configDigest[:2], uint16(ocrtypes.ConfigDigestPrefixLLO))
+	if !(configDigest[0] == 0 || configDigest[1] == 6) {
+		// assertion
+		panic("unexpected mismatch")
```

Comment:
Why panic so much? It looks like the callers have the ability to return errors.

Reply:
![dont-panic-hitchhikers-guide-to-the-galaxy](https://github.com/smartcontractkit/chainlink/assets/4147639/7c1f7ae2-5de9-4f2b-b758-b590baec708f)

Reply:
In all seriousness, there isn't a good reason. Many of these panics could indeed be errors.

Reply:
Removed some unnecessary panics